### ISSUE TITLE
Zooming in at certain positions with Orbit Controller doesnt work as expected

### ISSUE THREAD

**Issue by sharon92** [pygfx/pygfx#532]
If you run my example of Quad mesh:
https://github.com/sharon92/pygfx/blob/main/examples/other/MeshQt6.py

and try to zoom in from a low angle (I don't know how to describe this, but if you come in from a low elevation and try to zoom in)
![zooming_issue (1)](https://github.com/pygfx/pygfx/assets/51754340/eaa09bda-1336-4877-9afe-27d4265c841e)

i am printing the camera state on the left side and the width and height gets lower and lower but it doesnt change the view.

**Comment by almarklein** [comment]
The orbit controller revolved around a virtual center point. As you zoom in, it zooms in at that point. I think what happens is that as you're zooming in on that point (which is some meters above the riverbed), the landscape basically becomes the background.

I can see how this is confusing though 🤔  How would you expect zooming to work in this case?

**Comment by Korijn** [comment]
I would say #509 or the FlyController would be ideal to explore this kind of scene :)

**Comment by sharon92** [comment]
@almarklein yep exactly this! Maybe it is possible to update the center point as I move around the map, so its always in the middle of my updated scene.

**Comment by sharon92** [comment]
I asked chatgpt how would it go about keeping the camera aligned with ground level, and it said I need to point the camera up vector to normal of ground points.

I tried this inside `_update_rotate` of orbit control, supplying the vertices as ground_points variable to the controller
```
        if hasattr(self,'ground_points') and len(self.ground_points)>0:
            pts = []
            dst = []
            for geom in self.ground_points:
                dist = np.sum((geom.data[:,:-1] - position[:2])**2,axis=1)
                nearest_idx = np.argsort(dist)[:3]
                pts += [*geom.data[nearest_idx]]
                dst += [*dist[nearest_idx]]
            nearest_idx = np.argsort(dst)[:3]
            pts = np.array(pts)[nearest_idx]
            up = normals_from_vertices(pts,np.array([0,1,2])).mean(axis=1)
            print(up)
        else:
            up = camera_state["reference_up"]
```

Didn't work out quite well though and I got a headache :D

**Comment by almarklein** [comment]
The camera's reference_up is already the same as the surface normal. This is more a case of finding a suitable interaction method.

**Comment by sharon92** [comment]
> I think this center point is the source of confusion. The orbit controller has an implicit center/target. It rotates around that point. You can also zoom in/out, and then it still rotates around that same point. In other words, zooming does not change the center point by design.

@almarklein is there a way where I can visualize this center point in my scene?

**Comment by almarklein** [comment]
> is there a way where I can visualize this center point in my scene?

From the top of my head, this center point shoyld be at `camera.world.position + controller._get_target_vec(camera.get_state)`.

**Comment by Korijn** [comment]
> Would be great to know what you think. I would love to work on this, but I am not sure where to start and how to achieve the task.

I can explain to you how to implement all of these features in a new `MapController`.

I've been checking out [this video](https://www.youtube.com/watch?v=eyl8rt-vqlI) to understand better. Some things are not entirely clear yet, I'll ask you some questions.

> * Tilt the terrain (rotating it around a horizontal axis that goes through the center of the window) -> Orbit Controller but it rotates around an axis instead of a point
> * Rotate the terrain (around a vertical axis that goes through the center of the window) -> Orbit Controller but it rotates around an axis instead of a point

It looks to me like it works just like the orbit controller, can you explain what is different about rotating about an axis vs a point? It seems to me like the main difference is that the altitude (height) of the center of rotation is constrained.

> * Change the camera position (and the view center), moving it around in a horizontal plan -> pan controller in 3D

Based on the video, QGIS doesn't seem to take into account the height differences of the terrain; the camera just stays fixed to a flat plane, right?

> * Change the camera altitude

In the video I can see this is done by rotating, are there other controls available in QGIS to actually move the camera straight up and down?

> * Change the camera orientation (the camera is kept at its position but the view center point moves)

This is a big difference, I guess this is like how first person controls would allow you to change the camera orientation? Is it possible to actually look up? Above the horizon, I mean? This is the property of this control scheme that really sets it apart from the orbit controller.

> * Zoom in and out

Is zooming the same as with the orbit controller?

**Comment by sharon92** [comment]
> It looks to me like it works just like the orbit controller, can you explain what is different about rotating about an axis vs a point? It seems to me like the main difference is that the altitude (height) of the center of rotation is constrained.

@almarklein mentioned 
> The orbit controller has an implicit center/target. It rotates around that point. You can also zoom in/out, and then it still rotates around that same point. In other words, zooming does not change the center point by design.

QGIS uses the rotation around the horizontal and vertical axis that goes through center of the window, which means it's not a fixed axis. I could be anywhere in my terrain environment, and tilting and rotating the plane would work uniformly. Whereas with the orbit controller, I am limited to tilting and rotation within a certain radius of this mentioned implicit target.

> Based on the video, QGIS doesn't seem to take into account the height differences of the terrain; the camera just stays fixed to a flat plane, right?

I am not quite sure what you mean here. The elevation of the camera can be adjusted using (Page up and down)
https://github.com/pygfx/pygfx/assets/51754340/e0e3b94f-8c45-419b-a783-c4255d2b972f
https://github.com/pygfx/pygfx/assets/51754340/f7a77dbe-ce22-4519-813d-75a05fc8f3bb

> In the video I can see this is done by rotating, are there other controls available in QGIS to actually move the camera straight up and down?

Yes, in the current version, you can navigate up and down (stand/sit) using page up and down buttons.

> This is a big difference, I guess this is like how first person controls would allow you to change the camera orientation? Is it possible to actually look up? Above the horizon, I mean? This is the property of this control scheme that really sets it apart from the orbit controller.
> Let's say that you change the camera orientation, and look at the sky. Now where is the view center point located? Is the view center point some fixed distance away from the camera, or does it stay affixed to the "height zero plane" and prevent you from looking up over the horizon?

It allows looking down -90° and looking up +90° (observe the direction arrows on top right)
https://github.com/pygfx/pygfx/assets/51754340/29429df4-92e1-4837-acca-a49bbd361ee4
I can't answer the second question. If looking up over the horizon implies doing a 180° rotation of the head of a person (0° being looking straight, parallel to ground), no this is not possible in QGIS. I can only look straight up (perpendicular to the ground)

> Is zooming the same as with the orbit controller?

Zooming is similar to the pan controller. With the orbit controller, zooming is kind of limited in the direction of the implicit target.

**Comment by almarklein** [comment]
I'm currently working on too many different things already. I can have a look when I have a bit more bandwidth.
